lib/cocoapods/downloader/cache.rb

```diff
@@ -69,6 +69,93 @@ def cache_descriptors_per_pod
         end
       end
 
+      # Convenience method for acquiring a shared lock to safely read from the
+      # cache. See `Cache.lock` for more details.
+      #
+      # @param  [Pathname] location
+      #         the path to require a lock for.
+      #
+      # @param  [Proc] block
+      #         the block to execute inside the lock.
+      #
+      # @return [void]
+      #
+      def self.read_lock(location, block = Proc.new)
+        Cache.lock(location, File::LOCK_SH, block)
+      end
+
+      # Convenience method for acquiring an exclusive lock to safely write to
+      # the cache. See `Cache.lock` for more details.
+      #
+      # @param  [Pathname] location
+      #         the path to require a lock for.
+      #
+      # @param  [Proc] block
+      #         the block to execute inside the lock.
+      #
+      # @return [void]
+      #
+      def self.write_lock(location, block = Proc.new)
+        Cache.lock(location, File::LOCK_EX, block)
+      end
+
+      # Creates a .lock file at `location`, aquires a lock of type
+      # `lock_type`, checks that it is valid, and executes `block` while holding
+      # on to that lock. Afterwards, the .lock file is deleted, which is why
+      # validation of the lock is necessary, as you might have a lock on a file
+      # that doesn't exist on the filesystem anymore.
+      #
+      # @param  [Pathname] location
+      #         the path to require a lock for.
+      #
+      # @param  [locking_constant] lock_type
+      #         the type of lock, either exclusive (File::LOCK_EX) or shared
+      #         (File::LOCK_SH).
+      #
+      # @param  [Proc] block
+      #         the block to execute inside the lock.
+      #
+      # @return [void]
+      #
+      def self.lock(destination, lock_type, block = Proc.new)
```

Comment:
Is there really a point in using these APIs if no block is given? How about adding something like:

```suggestion
      def self.lock(destination, lock_type, block)
        raise ArgumentError unless block_given?
```

(And then also update the convenience methods accordingly.)

Reply:
Does `Proc.new` still work, even if there is no block passed? I would have assumed that it would throw an exception, but maybe not. I suppose it's preferable to explicitly throw an `ArgumentError ` than relying on the behaviour of `Proc.new`. 🤷‍♂️

Reply:
I see. Yeah I think for readers of the code, aside from runtime behaviour, it will be more clear to raise an `ArgumentError` yourself.

Reply:
I just tried it and the current behaviour is that if you don't pass a block you get the following error:

    tried to create Proc object without a block (ArgumentError)

To me it's pretty clear that my mistake was that I didn't pass a block. :)

Reply:
Absolutely, but you need to run that code to know that, it isn’t necessarily clear to anyone reading the code—like me 😄